packages/core/services/geocoder-service.ts

```diff
@@ -0,0 +1,41 @@
+import { Injectable } from '@angular/core';
+import { bindCallback, from, Observable, of, throwError } from 'rxjs';
+import { map, switchMap } from 'rxjs/operators';
+import { Geocoder, GeocoderRequest, GeocoderResult, GeocoderStatus } from './google-maps-types';
+import { MapsAPILoader } from './maps-api-loader/maps-api-loader';
+
+declare var google: any;
+
+@Injectable({ providedIn: 'root' })
+export class AgmGeocoder {
+  protected readonly geocoder$: Observable<Geocoder>;
+
+  constructor(loader: MapsAPILoader) {
+    this.geocoder$ = from(loader.load()).pipe(
```

Comment:
this is problematic, because since it's a cold observable, it ends up constructing Geocoder object on every geocode request. You should replace the whole thing with this:

```typescript
  constructor(loader: MapsAPILoader) {
    const connectableGeocoder$ = defer(() => new Observable(subscriber => {
      loader.load().then(() => subscriber.next());
    }))
    .pipe(
      map(() => this._createGeocoder()),
      multicast(new ReplaySubject(1)),
    ) as ConnectableObservable<Geocoder>;
    connectableGeocoder$.connect(); // ignore the subscription
    // since we will remain subscribed till application exits
    this.geocoder$ = connectableGeocoder$;
  }
```

Reply:
I've implemented this, though I didn't understand what I wrote. Would it be too much of a hassle to ask you to go through what's happening in this code?

Reply:
Sure.

Basically, Observables are created, and then send data down the pipeline to subscribers. Cold observers do the whole process from the start for every new subscriber, including initial creation. Hot observables create and send data without regard if there are subscribers or not.

The way you had it before was from(loader.load()). `from` is a function that creates a cold observable, which will call the `Promise.then` every time there is a subscriber ( and we subscribe every time we send a new geolocation request). THis isn't so bad, since a resolved promise sends its data immediately, but then the second step in the pipe is to create the geocoder. Since on a cold observable every step is done for every subscription, the geocoder will be created for every subscription (every request). 

The solution is to use multicasting. Multicasting creates a shared observable, which will does not resubscribe to its source for every new subscriber, but subscribes only once. Thus, all the steps in the pipe before the multicast will be done only once. 

When those operations from creation up to the multicast be pefrmored? Without multicast, they are performed on every subscription. With multicast they are performed manually, when you call `ConnectableObservable.connect` method. See, multicast operator doesn't return a regular observable, it returns a subclass, however typescript doesn`t recognize that pipe steps can return a different observable type, so we have to manually cast it. If we do not run `.connect` method, then subscribers to our multicasted observable will not receive any signals, since the upstream from multicast was never initialized.

The next problem is that on regular observable, creation and piping is done on every subscription, and data starts flowing for that subscriber only after subscribing. When multicast, subsription happens immediately after calling .connect, and data starts flowing. If a subscriber subscribes to the multicasted observable _after_ some data was already pushed, then this subscriber will never receive the data. This is true not just for multicasted observables, but any hot observable. In our case, the loader.load() method, and subsequent geocoder creation happened synchronously, as soon as we hit .connect(). If any subscriptions are added afterwards (meaning geocode requests sent), they will not receive the geocoder, and be left forever hanging. To solve this, we use a ReplaySubject in our multicast operator. ReplaySubject replays an `x` number of latest values on every subscription. Thus the new subscriptions that missed the values because they came too late, will still get the old data. We specifically told it to replay only one value, since we are only interested in one geocoder, and we know that upstream will only send the geocoder once. 

Finally, you originally converted loader.load promise to observable with a `from` function. The `from` function waits for the promise to resolve, after which it sends the promise's data down the pipeline, and _closes the observable_. This means that as soon as we connect with our multicast, we will get the load signal, create a geolocator, and then shut the whole pipeline down. If any subscribers subscribe to a closed observable, they will immediately receive the `closed` message, and nothing else. Even the ReplaySubject doesn't help in that case, since once the source is closed, it must close also. Therefore, we must create an observable ourselves, which will wait for the promise's value, and then **not** close.

While writing this, I realized we can slightly simplify the first few lines of the constructor like this:
```typescript
    const connectableGeocoder$ = new Observable(subscriber => {
      loader.load().then(() => subscriber.next());
    })
```

Reply:
Thank you very much, it's crystal clear.

I've added the latest simplification and pushed it to the branch.